File: coinz/app/src/main/java/eu/zerovector/coinz/SmartViewPager.java
package eu.zerovector.coinz;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.MotionEvent;

public class SmartViewPager extends ViewPager {

    public SmartViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    // We want to disable swiping to drag pages whilst on the map (which is page zero)
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (getCurrentItem() == 0) return false;
        else return true;
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {
        if (getCurrentItem() == 0) return false;
        else return true;
    }
}

File: coinz/app/src/main/java/eu/zerovector/coinz/SmartViewPager.kt
package eu.zerovector.coinz

import android.content.Context
import android.support.v4.view.ViewPager
import android.util.AttributeSet
import android.view.MotionEvent
import eu.zerovector.coinz.Data.bool


class SmartViewPager : ViewPager {
    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    private var oldX: Float = 0.0f
    private var oldY: Float = 0.0f

    // Only allow movement if it's vertical; otherwise allow everything normally.
    override fun onTouchEvent(event: MotionEvent): Boolean {
        return CheckDirsForbidHor(event) and super.onTouchEvent(event)
    }

    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
        return CheckDirsForbidHor(event) and super.onInterceptTouchEvent(event)
    }

    fun CheckDirsForbidHor(event: MotionEvent): bool {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                oldX = event.x
                oldY = event.y
            }
            MotionEvent.ACTION_MOVE -> {
                val newX = event.x
                val newY = event.y
                val deltaX = oldX - newX
                val deltaY = oldY - newY
                // Return "true" if the vertical delta is larger than the horizontal delta
                return Math.abs(deltaY) > Math.abs(deltaX)
            }
        }
        return true // just to be safe
    }
}